extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java

```diff
@@ -271,6 +287,61 @@ public void build(
                 ArcUndeclaredThrowableException.class.getName());
 
         resteasyServerConfig.produce(new ResteasyServerConfigBuildItem(path, resteasyInitParameters));
+
+        Set<DotName> autoInjectAnnotationNames = autoInjectAnnotations.stream().flatMap(a -> a.getAnnotationNames().stream())
+                .collect(Collectors.toSet());
+        annotationsTransformer.produce(new AnnotationsTransformerBuildItem(new AnnotationsTransformer() {
+
+            @Override
+            public boolean appliesTo(Kind kind) {
+                return kind == Kind.CLASS;
+            }
+
+            @Override
+            public void transform(TransformationContext transformationContext) {
+                ClassInfo clazz = transformationContext.getTarget().asClass();
+                if (clazz.classAnnotation(ResteasyDotNames.PATH) != null) {
+                    // Root resources - no need to add scope, @Path is a bean defining annotation
+                    if (clazz.classAnnotation(DotNames.TYPED) == null) {
+                        // Add @Typed(MyResource.class)
+                        transformationContext.transform().add(createTypedAnnotationInstance(clazz)).done();
+                    }
+                    return;
+                }
+                if (BuiltinScope.isIn(clazz.classAnnotations())) {
+                    // Skip classes annotated with built-in scope
+                    return;
+                }
+                if (clazz.classAnnotation(ResteasyDotNames.PROVIDER) != null) {
+                    Transformation transformation = null;
+                    if (clazz.annotations().containsKey(DotNames.INJECT)
+                            || hasAutoInjectAnnotation(autoInjectAnnotationNames, clazz)) {
```

Comment:
@gwenneg This should fix #4020 for any "auto-inject" annotation.

Reply:
@mkouba Thanks!